File: src/git/group/List/TList.java
package git.group.List;

import git.group.Builder.UserType;
import git.group.Comparator.Comparator;

import java.io.Serializable;
import java.util.ArrayList;

public class TList implements Serializable
{
	private class Node implements Serializable
	{
		public Node   next;
		public Object data;

		public Node(Object data)
		{
			this.data = data;
			this.next = null;
		}
	}

	private Node       head;
	private Node       tail;
	private int        size;
	private UserType builder;
	private Comparator comparator;

	public TList(UserType builder)
	{
		this.builder    = builder;
		this.comparator = builder.getTypeComparator();
		this.head       = null;
		this.tail       = null;
		this.size       = 0;
	}

	public TList(UserType[] arr) {
		for (int i=0; i<arr.length; i++)
		{
			pushEnd(arr[i]);
		}
	}

	public boolean clear()
	{
        if (head == null)
        {
            return false;
        }

        while (head != null)
        {
            delete(0);
        }

		return true;
	}

	public boolean pushFront(Object obj)
	{// vstavka v front

		Node nNode = new Node(obj);

		if (head == null)
		{
			head = nNode;
			tail = nNode;
		}
		else
		{
			Node temp = head;
			head      = nNode;
			head.next = temp;
		}
		size++;
		return true;
	}

	public boolean pushEnd(Object data)
	{
		Node nNode = new Node(data);

		if (head == null)
		{
			head = nNode;
			tail = nNode;
		}
		else
		{
			tail.next = nNode;
			tail      = tail.next;
		}
		size++;
		return true;
	}

	private void pushEnd(TList toInsert)
	{
		if (toInsert != null)
		{
			tail.next = toInsert.head;
			tail      = toInsert.tail;
			size += toInsert.size;
		}
	}

	public boolean add(Object data, int index)
	{
		Node nNode = new Node(data);

		if (head == null)
		{
			head = nNode;
			tail = nNode;
		}
		else
		{
			Node temp, current;
			temp    = head;
			current = null;

			for (int i = 0; i < index; i++)
			{
				current = temp;
				temp    = temp.next;
			}

			current.next = nNode;
			nNode.next   = temp;
		}
		size++;
		return true;
	}

	public boolean delete(int index)
	{
        if (size < 0 || index < 0)
        {
            return false;
        }

		Node toDel, toDelPrev = null;

		if (head == null)
		{
			System.out.println("List is empty");
			return false;
		}
		else
		{
			if (head != tail)
			{
				//Поиск ноды и её предщественника
                if (index > 0)
                {
                    toDelPrev = findNode(index - 1);
                    toDel     = toDelPrev.next;
                }
                else
                {
                    toDel = head;
                }

				if (toDelPrev != null)
				{
					toDelPrev.next = toDel.next;
					toDel          = null;
				}
				//Попали в голову
				else
				{
					head  = toDel.next;
					toDel = null;
				}
			}
			else
			{
				head = tail = null;
			}
		}
		size--;
		return true;
	}

	public Object find(int index)
	{
		Object dataNode;
		Node   current = head;

		if (index == 0)
		{
			dataNode = current.data;
			return dataNode;
		}

        for (int i = 0; i < index; i++)
        {
            current = current.next;
        }
		dataNode = current.data;
		return dataNode;
	}

	public int find(Object obj)
	{
		Node current = head;
		int  index   = 0;

        if (head == null)
        {
            return -1;
        }
        else
        {
            while (current != null)
            {
                if (current.data == obj)
                {
                    return index;
                }
                index++;
                current = current.next;
            }
        }
		return -1;
	}

	public void forEach(DoIt action) {
		ArrayList arr = new ArrayList();

		for (Node cur = head; cur != null; cur = cur.next)
			arr.add(cur.data);

		for (int i=0; i<arr.size();i++) {
			String str;
			if (arr.get(i) == null) str = "null ";
			else str = arr.get(i).toString() + " ";
			action.doIt(str);
		}
	}

	public TList sort()
	{
		TList r = quicksort(this);
		return r;
	}

	private TList quicksort(TList list)
	{
        if (list == null)
        {
            return list;
        }
		Node head_ = list.head;
		Node it    = head_.next;
        if (it == null)
        {
            return list;
        }
		TList lesser  = null;
		TList greater = null;
		while (it != null)
		{
			int comp = comparator.compare(it.data, head_.data);
			if (comp < 0 || comp == 0)
			{
                if (lesser == null)
                {
                    lesser = new TList(builder);
                }
				lesser.pushEnd(it.data);
			}
			else
			{
                if (greater == null)
                {
                    greater = new TList(builder);
                }
				greater.pushEnd(it.data);
			}
			it = it.next;
		}

		lesser  = quicksort(lesser);
		greater = quicksort(greater);

		TList buf = new TList(builder);
		buf.pushEnd(head_.data);
		buf.pushEnd(greater);
        if (lesser == null)
        {
            return buf;
        }
		lesser.pushEnd(buf);
		return lesser;
	}

	private Node findNode(int id)
	{
		Node res = head;
        for (int i = 0; i < id; i++)
        {
            res = res.next;
        }
		return res;
	}

	//GET
	//SET
	public int getSize()
	{
		return this.size;
	}

	public UserType getBuilder()
	{
		return builder;
	}

	public String toString() {
		Node cur = head;
		String str ="";
		for (int i = 0; i < size; i++) {
			str += (cur.data.toString());
			str+="\n";
			cur = cur.next;
			}
		return str;
	}

	public boolean setBuilder(UserType builder)
	{
		if (size == 0)
		{
			this.builder    = builder;
			this.comparator = builder.getTypeComparator();
			return true;
		}
		return false;
	}
}

File: src/git/group/scala/TList.scala
package git.group.scala

import git.group.java.List.{DoIt, TListInterface}
import git.group.java.Builder.UserType
import git.group.java.Comparator.Comparator

import java.util.function.Consumer
import scala.collection.mutable.ListBuffer

class TList() extends TListInterface
{
  class Node(data1:Any)
  {
    var data:Any = data1
    var next:Node = null
  }

  private var head:Node = null
  private var tail:Node = null
  private var size:Int = 0
  private var builder:UserType = null

  def this(build:UserType)
  {
    this()
    builder = build
  }

  def this(arr: Array[UserType])
    {
      this()
      for (i <- 0 until arr.length) {
        pushEnd(arr(i))
      }
    }

  def clear():Boolean=
  {
    if (head == null)
      return false

    while (head != null) delete(0)
    true
  }

  def pushFront(data:Any):Boolean =
  {
    var nNode:Node = new Node(data)

    if (head == null)
    {
      head = nNode
      tail = nNode
    }
    else
    {
      val tmp:Node = head


      head = nNode
      head.next = tmp
    }
    size = size + 1

    true
  }

  def pushEnd(data:Any):Boolean =
  {
    val nNode:Node = new Node(data)

    if (head == null){
      head = nNode
      tail = nNode
    }
    else {
      tail.next = nNode
      tail = nNode
    }
    size = size + 1
    true
  }

  def add(data:Any, index:Int):Boolean =
  {
    if (index == 0){
      return pushFront(data)
    }
    val nNode:Node = new Node(data)

    if (head == null) {
      head = nNode
      tail = nNode
    }
    else {
      var tmp: Node = head
      var current: Node = null
      var n: Int = 0
      while (n < index) {
        current = tmp
        tmp = tmp.next
        n = n + 1
      }
      current.next = nNode
      nNode.next = tmp
    }
    size = size + 1
    true
  }

  def delete(index:Int):Boolean = {
    if (index < 0) {
      return false
    }
    var toDel: Node = null
    var toDelPrev: Node = null

    if (head == null) {
      println("List is empty")
      return false
    }
    else {
      if (head != tail) {
        if (index > 0) {
          toDelPrev = findNode(index-1)
          toDel = toDelPrev.next
        }
        else toDel = head

        if (toDelPrev != null) {
          toDelPrev.next = toDel.next
          toDel = null
          if(toDelPrev.next == null)
            tail = toDelPrev
        }
        else {
          head = toDel.next
          toDel = null
        }

      }
      else {
        head = null
        tail = null
      }

    }
    size = size - 1
    true
  }

  private def findNode(id:Int):Node = {
    var res:Node = head
    var n:Int = 0
    while (n < id){
      res = res.next
      n = n + 1
    }
    res
  }

  def find(index:Int):Any = {
    var current:Node = head
    var dataNode:Any = current.data
    if (index == 0){
      dataNode = current.data
      return dataNode
    }
    var n:Int = 0
    while (n < index){
      current = current.next
      n = n + 1
    }
    dataNode = current.data
    dataNode
  }

  def sort2(xs: List[UserType], comparator: Comparator): List[UserType] ={
    if(xs.length <= 1) xs
    else{
      val pivot = xs(xs.length / 2)
      List.concat(
        sort2(xs.filter((x: UserType) => comparator.compare(pivot, x) > 0), comparator),
        xs.filter((x: UserType) => comparator.compare(pivot, x) == 0),
        sort2(xs.filter((x: UserType) => comparator.compare(pivot, x) < 0), comparator),
      )
    }
  }

  def sort(comparator: Comparator): TList ={
    var list : ListBuffer[UserType] = new ListBuffer[UserType]
    this.forEach((x : UserType) => list += x)
    var s_list = sort2(list.toList, comparator)
    var my_s = new TList()
    s_list.foreach((x: UserType) => my_s.pushEnd(x))
    return my_s;
  }

  def forEach(action: Consumer[UserType]): Unit = {
    var i:Int = 0
    var cur:Node = head
    while (i < size)
    {
      action.accept(cur.data.asInstanceOf[UserType])
      cur = cur.next
      i += 1
    }
  }

  def getSize:Int = size

  def getBuilder:UserType = builder

  override def toString: String = {
    var cur = head
    var str = ""
    for (i <- 0 until size) {
      str += cur.data.toString
      str += "\n"
      cur = cur.next
    }
    str
  }
}
